### ISSUE TITLE
重构后使用NPU训练GKD出现teacher server模式loss过大问题

### ISSUE THREAD

**Issue by jianingsong-cloud** [modelscope/ms-swift#9671]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new bug report. / 我已经搜索过现有的 issues，确认这是一个新的 bug report。

### Bug Description / Bug 描述

问题一：我在尝试使用ms-swift-4.4.0进行GKD训练时，验证发现使用teacher_model_api和teacher_model模式会在相同环境下有较大的loss差异（USE_MCORE_GDN=1）
使用teacher_server_api的初始loss:
{‘loss’: 0.23914044, ‘reward’: 2.46875, ‘grad_norm’: 0.185312, ‘learning_rate’: 0.0, ‘iteration’: ‘1/1389’, ‘elapsed_time’: ‘7m 41s’, ‘remaining_time’: ‘7d 9h 51m 55s’, ‘memory(GiB)’: 25.53, ‘train_speed(s/it)’: 461.322198}
Train: 0%|| 1/1389 [07:41<177:51:48, 461.32s/it]
使用teacher_model的初始loss:
{‘loss’: 0.0317606, ‘reward’: 2.59375, ‘grad_norm’: 0.12028129, ‘learning_rate’: 0.0, ‘iteration’: ‘1/1389’, ‘elapsed_time’: ‘9m 30s’, ‘remaining_time’: ‘9d 3h 50m 18s’, ‘memory(GiB)’: 18.85, ‘train_speed/s(it)’: 570.186143}
Train: 0%|| 1/1389 [09:30<219:50:11, 570.18s/it]

问题二：当使用teacher_model_api+padding_free为true时，会出现warning，并且loss进一步增大：（USE_MCORE_GDN=1）
[WARNINGSwift] The number of tokens returned by the teacher server differs from that of the trainer. This may be caused by non-aligned processing.
[WARNINGSwift] The number of tokens returned by the teacher server differs from that of the trainer. This may be caused by non-aligned processing.
[WARNINGSwift] The number of tokens returned by the teacher server differs from that of the trainer. This may be caused by non-aligned processing.
[WARNINGSwift] The number of tokens returned by the teacher server differs from that of the trainer. This may be caused by non-aligned processing.
[WARNINGSwift] The number of tokens returned by the teacher server differs from that of the trainer. This may be caused by non-aligned processing.
[WARNINGSwift] The number of tokens returned by the teacher server differs from that of the trainer. This may be caused by non-aligned processing.
{‘loss’: 0.60964227, ‘reward’: 2.375, ‘grad_norm’: 0.19536629, ‘learning_rate’: 0.0, ‘iteration’: ‘1/1389’, ‘elapsed_time’: ‘13m 14s’, ‘remaining_time’: ‘12d 18h 5m 52s’, ‘memory(GiB)’: 25.88, ‘train speed(s/it)’: 793.913224}
Train: 0% | 1/1389 [13:13<306:45, 793.91s/it]

问题三：USE_MCORE_GDN=1速度会很慢，USE_MCORE_GDN=0的速度会非常快，对比：
USE_MCORE_GDN=1
{‘loss’: 0.0317606, ‘reward’: 2.59375, ‘grad_norm’: 0.12028129, ‘learning_rate’: 0.0, ‘iteration’: ‘1/1389’, ‘elapsed_time’: ‘9m 30s’, ‘remaining_time’: ‘9d 3h 50m 18s’, ‘memory(GiB)’: 18.85, ‘train_speed/s(it)’: 570.186143}
Train: 0%|| 1/1389 [09:30<219:50:11, 570.18s/it]
USE_MCORE_GDN=0
{‘loss’: 0.0290553, ‘reward’: 2.59375, ‘grad_norm’: 0.20688038, ‘learning_rate’: 0.0, ‘iteration’: ‘1/1389’, ‘elapsed_time’: ‘2m 14s’, ‘remaining_time’: ‘2d 3h 44m 60s’, ‘memory(GiB)’: 14.38, ‘train_speed(s/it)’: 134.221884}
Train: 0% | 1/1389 [02:14<51:44:53, 134.22s/it]

问题四：当model_path Qwen3.5-9B 搭配teacher_model Qwen3.5-27B时，loss会非常大(USE_MCORE_GDN=0/1都有此情况)：

{'loss': 2.50792599, 'reward': 2.8828125, 'grad_norm': 19.99350166, 'learning_rate': 2e-05, 'iteration': '1/1736', 'elapsed_time': '19m 33s', 'remaining_time': '23d 13h 25m 27s', 'memory(GiB)': 16.83, 'train_speed(s/it)': 1173.214682}

### How to Reproduce / 如何复现

ms-swift版本：4.4.0dev0 commit id 4c1cce681
运行环境：Ascend 910B
transformers                             5.12.1
triton                                   3.5.0
triton_ascend                            3.2.1
flash-linear-attention                   0.5.1
mcore_bridge                             1.5.0.dev0 
mindspeed                                0.16.0

teacher_model_api shell：
model_path=/cache/Qwen3.5-9B/
export ASCEND_RT_VISIBLE_DEVICES=4,5
swift deploy \
    --model ${model_path} \
    --infer_backend vllm \
    --port 8001 \
    --max_logprobs 64 \
    --vllm_gpu_memory_utilization 0.85 \
    --vllm_max_model_len 4096 \
    --vllm_tensor_parallel_size 2 \
    --max_new_tokens 512 \
    --served_model_name Qwen3.5-9B

export ASCEND_RT_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
export NPROC_PER_NODE=4
export USE_MCORE_GDN=1
PYTORCH_NPU_ALLOC_CONF='expandable_segments:True' \
megatron rlhf \
    --rlhf_type gkd \
    --model ${model_path} \
    --teacher_model_server http://localhost:8001 \
    --gkd_logits_topk 64 \
    --offload_bridge true \
    --optimizer_cpu_offload true \
    --offload_teacher_model true \
    --enable_thinking false \
    --external_plugins examples/train/rlhf/opsd/opsd_plugin_vla.py \
    --dataset ${data_path} \
    --use_vllm true \
    --vllm_mode server \
    --vllm_server_base_url http://localhost:8000 \
    --tuner_type lora \
    --lora_rank 64 \
    --lora_alpha 128 \
    --lmbda 1.0 \
    --beta 0.5 \
    --temperature 1.2 \
    --sft_alpha 0 \
    --torch_dtype bfloat16 \
    --micro_batch_size 2 \
    --global_batch_size 16 \
    --train_iters 1000 \
    --lr 0 \
    --min_lr 0 \
    --save_steps 100 \
    --save_total_limit 10 \
    --logging_steps 1 \
    --max_length 3072 \
    --max_completion_length 512 \
    --tensor_model_parallel_size 4 \
    --pipeline_model_parallel_size 1 \
    --attention_backend flash \
    --finetune \
    --no_save_optim \
    --gradient_accumulation_fusion false \
    --no_save_rng \
    --val_dataset ${val_path} \
    --save_safetensors true \
    --context_parallel_size 1 \
    --num_train_epochs 4 \
    --freeze_vit true \
    --target_modules all-linear \
    --lr_warmup_fraction 0.05 \
    --clip_grad 1.0 \
    --top_p 0.95 \
    --top_k 50 \
    --repetition_penalty 1.0 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --padding_free false \
    --log_completions false \
    --eval_steps 1 \
    --lora_dropout 0.05 \
    --remove_unused_columns false \
    --packing false \
    --recompute_granularity full \
    --recompute_method uniform \
    --recompute_num_layers 4 \
    --output_dir ${output_dir}

teacher_model shell:
export ASCEND_RT_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
export NPROC_PER_NODE=4
export USE_MCORE_GDN=1
megatron rlhf \
    --rlhf_type gkd \
    --model ${model_path} \
    --teacher_model ${model_path} \
    --gkd_logits_topk 64 \
    --offload_bridge true \
    --optimizer_cpu_offload true \
    --offload_teacher_model true \
    --enable_thinking false \
    --external_plugins examples/train/rlhf/opsd/opsd_plugin_vla.py \
    --dataset ${data_path} \
    --use_vllm true \
    --vllm_mode server \
    --vllm_server_base_url http://localhost:8000 \
    --tuner_type lora \
    --lora_rank 64 \
    --lora_alpha 128 \
    --lmbda 1.0 \
    --beta 0.5 \
    --temperature 1.2 \
    --sft_alpha 0 \
    --torch_dtype bfloat16 \
    --micro_batch_size 2 \
    --global_batch_size 16 \
    --train_iters 1000 \
    --lr 0 \
    --min_lr 0 \
    --save_steps 100 \
    --save_total_limit 10 \
    --logging_steps 1 \
    --max_length 3072 \
    --max_completion_length 512 \
    --tensor_model_parallel_size 4 \
    --pipeline_model_parallel_size 1 \
    --attention_backend flash \
    --finetune \
    --no_save_optim \
    --gradient_accumulation_fusion false \
    --no_save_rng \
    --val_dataset ${val_path} \
    --save_safetensors true \
    --context_parallel_size 1 \
    --num_train_epochs 4 \
    --freeze_vit true \
    --target_modules all-linear \
    --lr_warmup_fraction 0.05 \
    --clip_grad 1.0 \
    --top_p 0.95 \
    --top_k 50 \
    --repetition_penalty 1.0 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --padding_free true \
    --log_completions false \
    --eval_steps 1 \
    --lora_dropout 0.05 \
    --remove_unused_columns false \
    --packing false \
    --recompute_granularity full \
    --recompute_method uniform \
    --recompute_num_layers 4 \
    --output_dir ${output_dir}

teacher_server_api+padding_free true:
model_path=/cache/Qwen3.5-9B/
export ASCEND_RT_VISIBLE_DEVICES=4,5
swift deploy \
    --model ${model_path} \
    --infer_backend vllm \
    --port 8001 \
    --max_logprobs 64 \
    --vllm_gpu_memory_utilization 0.85 \
    --vllm_max_model_len 4096 \
    --vllm_tensor_parallel_size 2 \
    --max_new_tokens 512 \
    --served_model_name Qwen3.5-9B

export ASCEND_RT_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
export NPROC_PER_NODE=4
export USE_MCORE_GDN=1
PYTORCH_NPU_ALLOC_CONF='expandable_segments:True' \
megatron rlhf \
    --rlhf_type gkd \
    --model ${model_path} \
    --teacher_model_server http://localhost:8001 \
    --gkd_logits_topk 64 \
    --offload_bridge true \
    --optimizer_cpu_offload true \
    --offload_teacher_model true \
    --enable_thinking false \
    --external_plugins examples/train/rlhf/opsd/opsd_plugin_vla.py \
    --dataset ${data_path} \
    --use_vllm true \
    --vllm_mode server \
    --vllm_server_base_url http://localhost:8000 \
    --tuner_type lora \
    --lora_rank 64 \
    --lora_alpha 128 \
    --lmbda 1.0 \
    --beta 0.5 \
    --temperature 1.2 \
    --sft_alpha 0 \
    --torch_dtype bfloat16 \
    --micro_batch_size 2 \
    --global_batch_size 16 \
    --train_iters 1000 \
    --lr 0 \
    --min_lr 0 \
    --save_steps 100 \
    --save_total_limit 10 \
    --logging_steps 1 \
    --max_length 3072 \
    --max_completion_length 512 \
    --tensor_model_parallel_size 4 \
    --pipeline_model_parallel_size 1 \
    --attention_backend flash \
    --finetune \
    --no_save_optim \
    --gradient_accumulation_fusion false \
    --no_save_rng \
    --val_dataset ${val_path} \
    --save_safetensors true \
    --context_parallel_size 1 \
    --num_train_epochs 4 \
    --freeze_vit true \
    --target_modules all-linear \
    --lr_warmup_fraction 0.05 \
    --clip_grad 1.0 \
    --top_p 0.95 \
    --top_k 50 \
    --repetition_penalty 1.0 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --padding_free true \
    --log_completions false \
    --eval_steps 1 \
    --lora_dropout 0.05 \
    --remove_unused_columns false \
    --packing false \
    --recompute_granularity full \
    --recompute_method uniform \
    --recompute_num_layers 4 \
    --output_dir ${output_dir}

### Additional Information / 补充信息

当前训练数据是包含图片的多模态数据，使用了OPSD的plugin

**Comment by hjh0119** [comment]
Issue 2 may have been fixed in https://github.com/modelscope/ms-swift/pull/9677

For the other issue, I am unable to reproduce it on my end. plz try again with latest code

**Comment by z0o0ey** [comment]
please retry with latest swift and vllm, feel free to create a new issue if it the problem persists.
